## Multiple Linear Regression

In [1]:
# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [2]:
#importing the dataset
dataset = pd.read_csv("Data.csv")

dataset.head(10)

R&D Spend  Administration  Marketing Spend       State     Profit
0  165349.20       136897.80        471784.10    New York  192261.83
1  162597.70       151377.59        443898.53  California  191792.06
2  153441.51       101145.55        407934.54     Florida  191050.39
3  144372.41       118671.85        383199.62    New York  182901.99
4  142107.34        91391.77        366168.42     Florida  166187.94
5  131876.90        99814.71        362861.36    New York  156991.12
6  134615.46       147198.87        127716.82  California  156122.51
7  130298.13       145530.06        323876.68     Florida  155752.60
8  120542.52       148718.95        311613.29    New York  152211.77
9  123334.88       108679.17        304981.62  California  149759.96

### Introduction the Dataset

On this dataset, it gives information about 50 startups' locations, profits and some related spends like R&D,  Markenting, Administration spends. Our case is searching the best parameters for decision of investing a startup. For instance, an investor wants to invest a startup but not sure about which spend parameters or location selection is the optimal criterions for the profit at the end of the day.

So, for this problem, we are using multiple linear regression model and after this process; investigating the best features according to the use some statistical parameters.

In [3]:
# Splitting the Dataset into the X, y

X = dataset.iloc[:,:-1].values  # [:,:-1] takes all features(independent variables) without last (-1) column 
y = dataset.iloc[:,4].values    # [:,4 ] takes last column(label, target, dependent variable)


In [4]:
#Encoding the independent variables
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X = LabelEncoder()
X[:,3] = labelencoder_X.fit_transform(X[:,3])
     
                            # X[:,3] takes just 3th order column where is "location"
                            # Also X[: -1] can be used for the selection of the last one.

            
onehotencoder = OneHotEncoder(categorical_features=[3]) 
X = onehotencoder.fit_transform(X).toarray()

                           # further explanation for OneHotEncoder, please look at the 
                           # 1-Data PreProcessing Part.
        

X[:5]   # get first 5 lists

array([[0.0000000e+00, 0.0000000e+00, 1.0000000e+00, 1.6534920e+05,
        1.3689780e+05, 4.7178410e+05],
       [1.0000000e+00, 0.0000000e+00, 0.0000000e+00, 1.6259770e+05,
        1.5137759e+05, 4.4389853e+05],
       [0.0000000e+00, 1.0000000e+00, 0.0000000e+00, 1.5344151e+05,
        1.0114555e+05, 4.0793454e+05],
       [0.0000000e+00, 0.0000000e+00, 1.0000000e+00, 1.4437241e+05,
        1.1867185e+05, 3.8319962e+05],
       [0.0000000e+00, 1.0000000e+00, 0.0000000e+00, 1.4210734e+05,
        9.1391770e+04, 3.6616842e+05]])

### Pay attention

Pay attention to the orders of columns are changed after onehotencoder process. So, we care new orders for further processes. 

Label Encoder encoded the locations as 0,1,2, after that, onehotencoder encoded that column according to the binarization.

In [5]:
#Avoiding the Dummy Variable Trap

X = X[:, 1:]   


X[:5]  # get first 5 lists

array([[0.0000000e+00, 1.0000000e+00, 1.6534920e+05, 1.3689780e+05,
        4.7178410e+05],
       [0.0000000e+00, 0.0000000e+00, 1.6259770e+05, 1.5137759e+05,
        4.4389853e+05],
       [1.0000000e+00, 0.0000000e+00, 1.5344151e+05, 1.0114555e+05,
        4.0793454e+05],
       [0.0000000e+00, 1.0000000e+00, 1.4437241e+05, 1.1867185e+05,
        3.8319962e+05],
       [1.0000000e+00, 0.0000000e+00, 1.4210734e+05, 9.1391770e+04,
        3.6616842e+05]])

### Why we are removing one of the dummy variables?

Assume that below equation has 2 categorical variable these are london(x4) and paris(x5) so we created two dummy variable to represent them.

#### y = b0 + x1b1 + x2b2 + x3b3 + x4b4 + x5b5 

So, naturally, if x4 is 0, london has to 1 or vice versa. Now let's think about our equation is below that;

#### y = b0 + x1b1 + x2b2 + x3b3 + x4b4 

Now, if london is 0, x4b4 equals zero, automatically, equation behaviour like paris is 1, because london is ineffective. It acts like electric switch ( if London is Off(0) --> Paris is On(1)  or if  Paris is Off(0) --> London is On(1).

So we assume that our core equation -without categorical variables- acts like one of the dummy variables is 1, others are 0's.

#### y = b0 + x1b1 + x2b2 + x3b3 

This core equation acts like Paris is 1. So,  after the creating dummy variables from categorical variables, we  should omit one dummy variable from the dataset. 

If there are two categorical sets on the dataset, like;

Location : London, Ankara, LA, New Mexico

Material : Metal, Wood, Plastic, Hybrid

After encoding the categorical variables, we should omit totally 2 dummy variables from each dummy sets.


In [6]:
# Splitting the dataset into the Training set and Test set        
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size = 0.2 ,random_state = 0)

In [7]:
#Fitting Multiple Linear Regression to Training sets

from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [8]:
# Predicting the Test set results.     
                                      
y_pred = regressor.predict(X_test)

"y_prediction", y_pred[:10] , "y_test_set", y_test[:10]

('y_prediction',
 array([103015.20159795, 132582.27760817, 132447.73845176,  71976.09851257,
        178537.48221058, 116161.24230165,  67851.69209675,  98791.73374686,
        113969.43533013, 167921.06569553]),
 'y_test_set',
 array([103282.38, 144259.4 , 146121.95,  77798.83, 191050.39, 105008.31,
         81229.06,  97483.56, 110352.25, 166187.94]))

There is not optimizing but the model predictions are not bad. For example, first observation in real case is 103282, prediction is 103015, this is not bad these difference are just 267.  

### Adding a column for X0 on the multiple linear equation 

y= b0 + b1.x1 + b2.x2 + b3.x3 + ... + bn.xn 

As we know, above equation is an example of multiple variabled equation and we executed our model with linear regressor model. On the learning model, regressor automatically creates b0 constant(intercept) variable, so there is no problem occur at the process.

As we now our dataset does not have b0.x0 column, because it just includes spends and location columns. Naturally, it is not hope for a dataset includes a x0 = 1 intercept column by itself.

Based on all this information, we want to improve model performance and decided to work with only most significant features. For this purpose, we trying to create a backward elimination model to eliminate these non-significant features. For this, statsmodels library will be used for observing some statistical parameters (like R-squared, p-value), but this library is not add x0.b0 by default. So, this x0 variable will be added by aid of numpy on the below. 

##### "An intercept is not included by default and should be added by the user." by OLS document

In [9]:
import statsmodels.formula.api as sm
X = np.append(arr = np.ones((len(X),1)).astype(int), values = X, axis = 1) # adds one column on the first order.

X[:5]

array([[1.0000000e+00, 0.0000000e+00, 1.0000000e+00, 1.6534920e+05,
        1.3689780e+05, 4.7178410e+05],
       [1.0000000e+00, 0.0000000e+00, 0.0000000e+00, 1.6259770e+05,
        1.5137759e+05, 4.4389853e+05],
       [1.0000000e+00, 1.0000000e+00, 0.0000000e+00, 1.5344151e+05,
        1.0114555e+05, 4.0793454e+05],
       [1.0000000e+00, 0.0000000e+00, 1.0000000e+00, 1.4437241e+05,
        1.1867185e+05, 3.8319962e+05],
       [1.0000000e+00, 1.0000000e+00, 0.0000000e+00, 1.4210734e+05,
        9.1391770e+04, 3.6616842e+05]])

### Backward Elimination  Model

Backward Elimination model is a process for eliminating the nonsignificant features. According to the backward elimation model, we control the p-value for each features. If p-value is greater than threshold value, it says usually that independent variable does not have more effect for the result or target value. If the p-value is smaller than the threshold value or significance level, it says generally that independent variable has a significant effect for target value.

In statistics, on the smaller p-value case; H0 hypothesis is rejected, because the variable is significant, on the greater p-value than threshold value; H0 is not refected and the that independent variable is nonsignificant.


#### Backward Elimination Model's Steps:

#1 - Selecting a significant level to stay in the model ( e.g. SL = 0.05)

#2 - Fit the all model with all possible predictors(independent variables)

#3 - Consider the all predictors with the highest p-value. If p-value > Significance Level, go to the Step 4, otherwise - > Model is Ready 

#4 Remove the Predictor

#5 Fit the model without eliminated variable, turn the 3th Step.


In [10]:
# Building the optional model using Backward Elimination

## Manuel Solution:
 
feature = [0,1,2,3,4,5]  #created a list for all features in the beginning
X_opt = X[:, feature]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit() # created regressor_OLS object 
regressor_OLS.summary()      # get the statistical summary

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.951
Model:                            OLS   Adj. R-squared:                  0.945
Method:                 Least Squares   F-statistic:                     169.9
Date:                Fri, 01 Jun 2018   Prob (F-statistic):           1.34e-27
Time:                        00:00:36   Log-Likelihood:                -525.38
No. Observations:                  50   AIC:                             1063.
Df Residuals:                      44   BIC:                             1074.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       5.013e+04   6884.820      7.281      0.000    3.62e+04     6.4e+04
x1           198.7888   3371.007      0.059      0.953   -6595.030    6992.607
x2           -41.8870   3256.039     -0.013      0.990   -6604.003    6520.229
x3             0.8060      0.046     17.369      0.000       0.712       0.900
x4            -0.0270      0.052     -0.517      0.608      -0.132       0.078
x5             0.0270      0.017      1.574      0.123      -0.008       0.062
==============================================================================
Omnibus:                       14.782   Durbin-Watson:                   1.283
Prob(Omnibus):                  0.001   Jarque-Bera (JB):               21.266
Skew:                          -0.948   Prob(JB):                     2.41e-05
Kurtosis:                       5.572   Cond. No.                     1.45e+06
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.45e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

As we know these independents variables are from the beginning: constant variable, x1 and x2 dummy variables, x3: R&D Spend, x4: Administration, x5:Marketing Spend. 

According the p-value strategy, we should eliminate the greatest P>|t| value from our dataset. We see that x2 (second dummy variable) has the greatest p-value, so it can be sat that has least significant effect for the target value. Thus we eliminate that column.  

In [11]:
feature = [0,1,3,4,5]   # variable x2 is eliminated.
X_opt = X[:, feature]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary() # getting the new parameters

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.951
Model:                            OLS   Adj. R-squared:                  0.946
Method:                 Least Squares   F-statistic:                     217.2
Date:                Fri, 01 Jun 2018   Prob (F-statistic):           8.49e-29
Time:                        00:00:36   Log-Likelihood:                -525.38
No. Observations:                  50   AIC:                             1061.
Df Residuals:                      45   BIC:                             1070.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       5.011e+04   6647.870      7.537      0.000    3.67e+04    6.35e+04
x1           220.1585   2900.536      0.076      0.940   -5621.821    6062.138
x2             0.8060      0.046     17.606      0.000       0.714       0.898
x3            -0.0270      0.052     -0.523      0.604      -0.131       0.077
x4             0.0270      0.017      1.592      0.118      -0.007       0.061
==============================================================================
Omnibus:                       14.758   Durbin-Watson:                   1.282
Prob(Omnibus):                  0.001   Jarque-Bera (JB):               21.172
Skew:                          -0.948   Prob(JB):                     2.53e-05
Kurtosis:                       5.563   Cond. No.                     1.40e+06
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.4e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

As we see, x1(first dummy variable) has greatest p-value and  has nonsignificant effect. So, we can eliminate it at this stage.

In [12]:
feature = [0,3,4,5]    # variable x1 is eliminated.
X_opt = X[:, feature]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary() # getting the new parameters

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.951
Model:                            OLS   Adj. R-squared:                  0.948
Method:                 Least Squares   F-statistic:                     296.0
Date:                Fri, 01 Jun 2018   Prob (F-statistic):           4.53e-30
Time:                        00:00:37   Log-Likelihood:                -525.39
No. Observations:                  50   AIC:                             1059.
Df Residuals:                      46   BIC:                             1066.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       5.012e+04   6572.353      7.626      0.000    3.69e+04    6.34e+04
x1             0.8057      0.045     17.846      0.000       0.715       0.897
x2            -0.0268      0.051     -0.526      0.602      -0.130       0.076
x3             0.0272      0.016      1.655      0.105      -0.006       0.060
==============================================================================
Omnibus:                       14.838   Durbin-Watson:                   1.282
Prob(Omnibus):                  0.001   Jarque-Bera (JB):               21.442
Skew:                          -0.949   Prob(JB):                     2.21e-05
Kurtosis:                       5.586   Cond. No.                     1.40e+06
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.4e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

As you see, x2 has the greatest p-value we should eliminate them but there is no called x2 variable our feature list. Be careful this case, because OLS summary process shows us the variable results after rescaled. So, if we eliminate the x2 variable, we should count its order ( constant: 0, x1: 1, x2: 2 --> 2nd order) after that, we find the the value for feature list according to this order. feature = [0,3,4,5].  feature[2] = 4. Thus, we can eliminate variable 4.

In [13]:
feature = [0,3,5]
X_opt = X[:, feature]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.950
Model:                            OLS   Adj. R-squared:                  0.948
Method:                 Least Squares   F-statistic:                     450.8
Date:                Fri, 01 Jun 2018   Prob (F-statistic):           2.16e-31
Time:                        00:00:37   Log-Likelihood:                -525.54
No. Observations:                  50   AIC:                             1057.
Df Residuals:                      47   BIC:                             1063.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       4.698e+04   2689.933     17.464      0.000    4.16e+04    5.24e+04
x1             0.7966      0.041     19.266      0.000       0.713       0.880
x2             0.0299      0.016      1.927      0.060      -0.001       0.061
==============================================================================
Omnibus:                       14.677   Durbin-Watson:                   1.257
Prob(Omnibus):                  0.001   Jarque-Bera (JB):               21.161
Skew:                          -0.939   Prob(JB):                     2.54e-05
Kurtosis:                       5.575   Cond. No.                     5.32e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 5.32e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Now, we notice that x2 has the greatest p-value (p-value(x2) > 0.05). So we should eliminate this ordered feature from our list. This is the variable 5.

In [14]:
feature = [0,3]
X_opt = X[:, feature]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.947
Model:                            OLS   Adj. R-squared:                  0.945
Method:                 Least Squares   F-statistic:                     849.8
Date:                Fri, 01 Jun 2018   Prob (F-statistic):           3.50e-32
Time:                        00:00:38   Log-Likelihood:                -527.44
No. Observations:                  50   AIC:                             1059.
Df Residuals:                      48   BIC:                             1063.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       4.903e+04   2537.897     19.320      0.000    4.39e+04    5.41e+04
x1             0.8543      0.029     29.151      0.000       0.795       0.913
==============================================================================
Omnibus:                       13.727   Durbin-Watson:                   1.116
Prob(Omnibus):                  0.001   Jarque-Bera (JB):               18.536
Skew:                          -0.911   Prob(JB):                     9.44e-05
Kurtosis:                       5.361   Cond. No.                     1.65e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.65e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

And we reach the end of the backward elimination model. Because the rest variables have the lower p-value from the significance level. So we can say that the fit model can be more precise for 3th feature that is R&D spend. 0st variable is the constant variable, this is already included on the fitting model by default.  

In [15]:
#Automated Backward Elimination Model

feature = [0,1,2,3,4,5]     # list of features
SL = 0.05                   # significance level
 
for num in range(0,len(feature)):                  
    X_opt = X[:, feature]
    regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
    order=0
    if max(regressor_OLS.pvalues) > SL:
        for i in range(0,len(regressor_OLS.pvalues)):
            if regressor_OLS.pvalues[i] == max(regressor_OLS.pvalues):
                feature.remove(feature[i])        
    else:
        continue           
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.947
Model:                            OLS   Adj. R-squared:                  0.945
Method:                 Least Squares   F-statistic:                     849.8
Date:                Fri, 01 Jun 2018   Prob (F-statistic):           3.50e-32
Time:                        00:00:42   Log-Likelihood:                -527.44
No. Observations:                  50   AIC:                             1059.
Df Residuals:                      48   BIC:                             1063.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       4.903e+04   2537.897     19.320      0.000    4.39e+04    5.41e+04
x1             0.8543      0.029     29.151      0.000       0.795       0.913
==============================================================================
Omnibus:                       13.727   Durbin-Watson:                   1.116
Prob(Omnibus):                  0.001   Jarque-Bera (JB):               18.536
Skew:                          -0.911   Prob(JB):                     9.44e-05
Kurtosis:                       5.361   Cond. No.                     1.65e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.65e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

So, automated process calculated the result very quickly.